src/coreclr/gc/objecthandle.cpp

```diff
@@ -1065,7 +1065,9 @@ void Ref_TraceNormalRoots(uint32_t condemned, uint32_t maxgen, ScanContext* sc,
     // promote objects pointed to by variable handles whose dynamic type is VHT_STRONG
     TraceVariableHandles(PromoteObject, uintptr_t(sc), uintptr_t(fn), VHT_STRONG, condemned, maxgen, flags);
 
-#if defined(FEATURE_COMINTEROP) || defined(FEATURE_REDHAWK)
+    GCToEEInterface::RefCountedHandleCallbacksBefore();
```

Comment:
This does not need to be called when `sc->concurrent == true`. Same applies to the "After" part.
When handles are scanned in the background (concurrently with managed threads), ref-counted handles are not scanned, so callbacks will be redundant.
See the condition at the line 1072.

Reply:
What kind of code is runs in this Begin/After callback? Do we have an example?

Reply:
@VSadov Thanks for the clarity. I may have misunderstood @Maoni0 when this location was suggested, will let her clarify but I applied your suggestion and updated the naming on the interop side.

>What kind of code runs in this Begin/After callback? Do we have an example?

Sure. An example is in the test which isn't how the product will do this but the idea is to let the other system know when those handles are being examined and tracking that state.

```cpp
    std::atomic_bool state = { false };

     void STDMETHODCALLTYPE BeginEndCb()
    {
#ifdef REF_TRACKER_DEBUG_CALLBACKS
        ::printf("BeginEndCb: %s -> %s\n",
            state ? "true" : "false",
            !state ? "true" : "false");
#endif // REF_TRACKER_DEBUG_CALLBACKS

        state = !state;
    }
```

Reply:
My concern is that on server GC this will end up being called N times in parallel where N is the number of heaps, which is typically #proc. So the `BeginEnd` callback will see N begin calls and N end calls in some random order. It looks like your callback is not expecting such scenario.

It could get worse if the real callback takes locks as that would have serializing effect on GC - it will force the handle marking to happen serially, one heap after another.

I think these Begin/End calls should happen only once per GC (background or regular) -
-  one call before marking handles and one after in a regular (nonconcurrent) mark, and 
-  one call before marking handles and one after in a nonconcurrent mark phase of a background GC. 

In the second case the callbacks should happen when worker threads are joined, so that only one worker thread makes calls, not all of them. 
I think we have convenient locations in GC to put these callbacks, but you should check with @Maoni0 if that makes sense and where exactly the callbacks should be placed.